ddht/v5_1/alexandria/partials/proof.py

```diff
@@ -29,6 +36,84 @@ def __str__(self) -> str:
         return f"{display_path(self.path)}: {self.value.hex()}"
 
 
+@to_tuple
+def get_padding_elements(
+    start_index: int, num_padding_chunks: int, path_bit_length: int
+) -> Iterable[ProofElement]:
+    """
+    Get the padding elements for a proof.
+
+    By decomposing the number of chunks needed into the powers of two which
+    make up the number we can construct the minimal right hand sub-tree(s)
+    needed to pad the hash tree.
+    """
+    for power_of_two in decompose_into_powers_of_two(num_padding_chunks):
+        depth = power_of_two.bit_length() - 1
+        left_index = start_index + power_of_two
+        left_path = chunk_index_to_path(left_index, path_bit_length)
+        padding_hash_tree_root = ZERO_HASHES[depth]
+        yield ProofElement(left_path[: path_bit_length - depth], padding_hash_tree_root)
+
+
+class DataPartial:
+    """
+    A wrapper around a partial data proof which allows data retrieval by
+    indexing or slicing.
+
+    Raise `IndexError` if the requested data is not part of the proof.
+    """
+
+    def __init__(self, length: int, segments: Tuple[Tuple[int, bytes], ...]) -> None:
+        self._length = length
+        self._segments = segments
+
+    def __len__(self) -> int:
+        return self._length
+
+    @overload
+    def __getitem__(self, index_or_slice: int) -> int:
+        ...
```

Comment:
Python must be so strange for anybody learning it from scratch today :sweat_smile:

Reply:
^ lol, not too bad in my experience. As someone not familiar with all the conventions: are we using `...` to mean `TODO` or `noop`?